File: src/main/java/com/gahyeonbot/adapters/unreal/IncrementalSentenceAccumulator.java
package com.gahyeonbot.adapters.unreal;

import java.util.ArrayList;
import java.util.List;

/** Incrementally emits speakable sentences without waiting for the full LLM response. */
public final class IncrementalSentenceAccumulator {
    private final int maxCharacters;
    private final StringBuilder pending = new StringBuilder();

    public IncrementalSentenceAccumulator(int maxCharacters) {
        if (maxCharacters < 20) throw new IllegalArgumentException("maxCharacters must be at least 20");
        this.maxCharacters = maxCharacters;
    }

    public List<String> accept(String delta) {
        if (delta == null || delta.isEmpty()) return List.of();
        pending.append(delta);
        return drain(false);
    }

    public List<String> finish() {
        return drain(true);
    }

    private List<String> drain(boolean finish) {
        List<String> ready = new ArrayList<>();
        int start = 0;
        for (int index = 0; index < pending.length(); index++) {
            char value = pending.charAt(index);
            int length = index - start + 1;
            if (sentenceEnd(value)
                    || length >= maxCharacters && safeBreak(value)
                    || length >= maxCharacters * 2) {
                add(ready, pending.substring(start, index + 1));
                start = index + 1;
            }
        }
        if (start > 0) pending.delete(0, start);
        if (finish && !pending.isEmpty()) {
            add(ready, pending.toString());
            pending.setLength(0);
        }
        return List.copyOf(ready);
    }

    private boolean sentenceEnd(char value) {
        return value == '.' || value == '?' || value == '!'
                || value == '。' || value == '？' || value == '！' || value == '\n';
    }

    private boolean safeBreak(char value) {
        return Character.isWhitespace(value) || value == ',' || value == '，';
    }

    private void add(List<String> ready, String value) {
        String normalized = value.trim();
        if (!normalized.isEmpty()) ready.add(normalized);
    }
}

File: desktop/src/conversation/incremental-sentence-accumulator.ts
export class IncrementalSentenceAccumulator {
  private pending = ''

  constructor(private readonly maxCharacters = 180) {
    if (maxCharacters < 20) throw new Error('maxCharacters must be at least 20')
  }

  accept(delta: string) {
    if (!delta) return []
    this.pending += delta
    return this.drain(false)
  }

  finish() {
    return this.drain(true)
  }

  private drain(finish: boolean) {
    const ready: string[] = []
    let start = 0
    for (let index = 0; index < this.pending.length; index++) {
      const value = this.pending[index]
      const length = index - start + 1
      const sentenceBoundary = /[.?!。？！\n]/u.test(value)
      const hardBoundary = length >= this.maxCharacters && !isHighSurrogate(value)
      if (sentenceBoundary || hardBoundary) {
        this.add(ready, this.pending.slice(start, index + 1))
        start = index + 1
      }
    }
    if (start > 0) this.pending = this.pending.slice(start)
    if (finish && this.pending) {
      this.add(ready, this.pending)
      this.pending = ''
    }
    return ready
  }

  private add(target: string[], value: string) {
    const normalized = value.trim()
    if (normalized) target.push(normalized)
  }
}

function isHighSurrogate(value: string) {
  if (!value) return false
  const code = value.charCodeAt(0)
  return code >= 0xD800 && code <= 0xDBFF
}

export function unseenFinalText(finalText: string, streamedText: string) {
  return finalText.startsWith(streamedText) ? finalText.slice(streamedText.length) : ''
}
